src/relayclient/RelayClient.ts

```diff
@@ -80,8 +81,38 @@ export class RelayClient {
     configOverride: Partial<GSNConfig>,
     overrideDependencies?: Partial<GSNDependencies>
   ) {
-    this.config = configureGSN(configOverride)
-    const dependencies = getDependencies(this.config, provider, overrideDependencies)
+    this.partialConfig = { provider, configOverride, overrideDependencies }
+    this.logger = console
+  }
+
+  initializingPromise?: Promise<void>
+
+  async init (): Promise<this> {
+    if (this.initializingPromise != null) {
+      await this.initializingPromise
+    }
+    if (this.initialized) {
+      return this
+    }
+    let initializingResolve!: () => void
+    this.initializingPromise = new Promise((resolve) => {
```

Comment:
would this work?

```
async init (): Promise<this> {
  if (this.initialized) {
      throw Error('Do not call init twice!')
    }
  this.initializingPromise = this._initInternal() // no await here
  // es-lint will be angry but this is actually functionally equivalent
  return this.initializingPromise
}

async relayTransaction (gsnTransactionDetails: GsnTransactionDetails): Promise<RelayingResult> {
  if (!this.initialized) {
    console.log('Suggestion here)
    await this.initializingPromise
  }
```


I feel like the trick with extracting the resolve of an artificial promise is a bit of a clutch. Not sure this is functionally equivalent though.

Reply:
looks good. cleaner than my construct.